abi.html

```diff
@@ -5827,6 +5957,298 @@ <h5><a href="#mangling.declaration-reference">5.1.6.2 References to declared ent
 </pre></code>
 </p>
 
+
+<a name="mangling.constant-values">
+<h5><a href="#mangling.constant-values">5.1.6.3 Constant values</a></h5>
+
+<p>
+It is sometimes necessary to mangle a constant value resulting from
+constant evaluation.  Currently this is limited to several places
+in the mangling of a non-dependent <a href="#mangle.template-arg">non-type
+template argument</a>.  In general, this is done by mangling a notional
+expression with the value and type of the constant.  For various reasons,
+these mangled expressions do not always correspond to valid source
+expressions.
+
+<p>
+The mangling of constant values is sensitive to the converted type of
+the constant.  For example, the expression <code>5</code> has type
+<code>int</code>, but when it is used as the argument for a template
+parameter of type <code>unsigned short</code>, it is converted to that
+type and mangled as <code>Lt5E</code>.  Some contexts require
+<i>precise typing</i>, meaning that constants must be mangled in a way
+that includes the exact converted type.  In other contexts, some
+information not necessary to distinguish constants is omitted in
+order to reduce symbol sizes and preserve existing manglings.
+The places in this ABI that call for the mangling of a constant value
+all indicate whether precise typing is required for a particular
+constant.
+
+<p>
+Values of integer, enumerated, floating-point, or complex type are
+mangled as <a href="#mangling.literal">literals</a> of the approproiate
+type.  Note that this can produce combinations which cannot normally be
+written in the source language, such as a literal of <code>short</code>
+type.
+
+<p>
+Null pointers and member pointers are mangled as as a literal
+<code>0</code> of their type.  For example:
+
+<code><pre>
+struct A;
+template &lt;void (A::*)()&gt; void f();
+
+f&lt;nullptr&gt; // _Z1fILM1AFvvE0EEvv
+</pre></code>
+
+<p>
+Non-null values of reference or pointer type are mangled using the
+<a href="#mangling.constant-object-reference">constant object reference
+expression</a> for the argument.  Note that, if this is a simple
+declaration reference, it may end up being an
+<code>&lt;<a href="#mangle.expr-primary">expr-primary</a>&gt;</code>
+and must be mangled without <code class=mangle>X...E</code> when
+mangled as a <code>template-arg</code>.
+
+<p>
+Non-type template argument values of member pointer type
+that are not null are mangled using the
+<a href="#mangling.member-pointer-reference">member pointer
+reference expression</a> for the argument.
+
+<p>
+Non-type template argument values of class type are mangled as
+a direct initialization (<code>tl</code>) of the class type
```

Comment:
I think we should use `il` instead of `tl <type>` when precise typing is not required. This would make a difference when mangling a union member that is itself of class type.

Reply:
That would be a total break with the provisional manglings implementations have been using for class template arguments.

Reply:
Hm, I suppose this would change all manglings for (eg) class templates with non-type parameters of class type. That said, the addition of flattening is a pretty big change too. How big a break are we prepared to accept here? Unlike some of the other manglings we're changing in this PR, these ones only affect C++20 onwards.